### PR TITLE
[dagster-airlift][sensor] make sensor implementation pluggable at top level

### PR DESCRIPTION
**Pull request by dpeng817**
## Summary & Motivation
Introduce a pluggability point for the sensor. Only bit that has changed from offline discussion is that I added a context argument so that users can perform their own logging.

## How I Tested These Changes
Added a new test for pluggable implementation.
Additional testing I would like to do:
- ensure that an error in pluggable fxn doesn't advance the cursor
- ensure that we handle errors in the pluggable function well.
- ensure that we gracefully error when effective timestamp metadata is missing.

## Changelog
NOCHANGELOG

### CODE REVIEW

**Comment by dpeng817** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25052?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#25053** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25053?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25052** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25052?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#25051** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25051?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#25050** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/25050?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @dpeng817 and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py [hunk 1]
```diff
@@ -132,13 +132,18 @@ def airflow_dag_sensor(context: SensorEvaluationContext) -> SensorResult:
                 end_date_lte=None,
                 dag_query_offset=0,
             )
+        updated_asset_events = (
+            list(event_transformer_fn(context, airflow_data, all_asset_events))
+            if event_transformer_fn
+            else all_asset_events
+        )
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py:139]: as we are calling into user code we should be far more paranoid with the returned values use a `user_code_error_boundary`

**Review by schrockn — CHANGES_REQUESTED** [review]
Let's put some structure in the invocation back to user space. At a minimum we should introspect on the return values and provide a clear error message.

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py (hunk 1 above)
Comment (schrockn) [reply examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py:139]: One could argue that the `user_code_error_boundary` might be overkill, as the sensor is already in "user space". But at a minimum introspect on return values out of `event_transformer_fn`

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py [hunk 2]
```diff
@@ -132,13 +144,36 @@ def airflow_dag_sensor(context: SensorEvaluationContext) -> SensorResult:
...
+                if not isinstance(
+                    asset_event, (AssetMaterialization, AssetObservation, AssetCheckEvaluation)
+                ):
+                    raise DagsterInvariantViolationError(
+                        f"Event transformer function must return AssetMaterialization, AssetObservation, or AssetCheckEvaluation objects. Got {type(asset_event)}."
+                    )
+                if EFFECTIVE_TIMESTAMP_METADATA_KEY not in asset_event.metadata:
+                    raise DagsterInvariantViolationError(
+                        f"All returned events must have an effective timestamp, but {asset_event} does not. An effective timestamp can be used by setting dagster_airlift.constants.EFFECTIVE_TIMESTAMP_METADATA_KEY with a dagster.TimestampMetadataValue."
+                    )
```
Comment (schrockn) [inline examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py:166]: probably worth extracting all this slop into a sep function to clean up the core loop

**Review by schrockn — APPROVED** [review]
Cool thanks.

File: examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py (hunk 2 above)
Comment (dpeng817) [reply examples/experimental/dagster-airlift/dagster_airlift/core/sensor/builder.py]: slop: extracted

**Pull request merged**